---
title: "Monocytes expression profile of children vs adults before and after malaria challenge"
author: "Z. Pava"
date: '2023-14-08'
output:
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
#Analysis:
##Differential expression analysis of bulk RNAseq by generalized linear mixed model using ##glmmseq package. 

#Descriptions:
##glmmseq calculates DEGs using raw counts data (i.e. HTSeq data). glmmseq allows model to ##include random effects, like patient ID and analyse longitudinal data. 
##It doesn't account for repeated measurements (2+)

```{r, libraries, echo=FALSE, results='hide',include= FALSE, warn.conflicts=F, quietly=T}
.libPaths("/Users/zuleip/miniconda3/envs/rnaseq-age/lib/R/library")

library(glmmSeq)
library(limma)
library(edgeR)
library(tidyverse)
library(kableExtra)
library(DESeq2)
library(tximeta)
```
**Input file format:** Raw counts file with gene ids as row names and libraries name as columns. #Multiple gene ids as extra columns (optional)

**note:**See script "data_exploration.R" to reproduce clean & formatted input files (i.e.counts and metadata)

```{r, input data, echo=TRUE, results='hide', warn.conflicts=F, quietly=T}
counts = read.delim(file = "../data/counts_Age_monos.txt", sep = "\t",
                        header=T, row.names=1)
meta = read.delim(file = "../data/metadata_Age_monos.txt", sep = "\t",
                      header=T, row.names=1)
```
**Estimating dispersion:**Using negative binomial models requires gene dispersion estimates to be made. Here we calculate dispersion with DESeq. edger and other methods can be used as well. 
```{r, estimating dispersion, echo=TRUE, results='hide', warn.conflicts=F, quietly=T}
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = meta,
                              design = ~state_bool*age_bool)
dds <- DESeq(dds)

dispersions <- setNames(dispersions(dds), rownames(counts))

res <- results(dds)
res <- res[order(res$padj),]
sum(res$padj < 0.05, na.rm=TRUE)

##Estimating size factors 
sizeFactors <- estimateSizeFactorsForMatrix(counts)
```
**Interpreting dispersion plots**DESeq2 dispersion estimates are inversely related to the mean and directly related to variance. Based on this relationship, the dispersion is higher for small mean counts and lower for large mean counts.
```{r plotting dispersion, include=TRUE}
plotDispEsts(dds, main="Monos Dispersal estimates_DESeq2")
```
**Interpreting an MA-plot**An MA-plot is a scatter plot of log2 fold changes (on the y-axis) versus the mean of normalized counts (on the x-axis). The MA-plot shows the log2 fold changes attributable to a given variable over the mean of normalized counts for all the samples in the DESeqDataSet. Points will be colored blue if the adjusted p value is less than 0.1. Points which fall out of the window are plotted as open triangles pointing either up or down.
```{r, MA plots, include=TRUE}
plotMA(res, ylim=c(-2,2), main="Monos MA Plot_DESeq2")
```
#Running the model and calculating q adjusted values.
```{r, running the model, echo=TRUE, results='hide', warn.conflicts=F, quietly=T}
##Running the model state accounting for age status
glmm_results <- glmmSeq(~state_bool*age_bool + (1 | samples),
                        countdata = counts,
                        metadata = meta,
                        dispersion = dispersions,
                        progress=TRUE,
                        cores = 4)
##Calculating q-values which are similar to p-adjusted values
glmm_results <- glmmQvals(glmm_results, cutoff = 0.05, verbose = TRUE)  
```
**RESULTS**
Getting data for plots and tables
```{r, getting data for plots and tables, echo=TRUE}
stats = as.data.frame(glmm_results@stats)
predict = as.data.frame(glmm_results@predict)
stats$gene.id <- rownames(stats)
predict$gene.id <- rownames(predict)

plotdata <- left_join(predict, stats, by = "gene.id")

plotdata1 = plotdata %>% mutate(
  #State_AdultFC=y_Adultstim-y_Adultunstim)
  State_AdultFC = log2(plotdata[, 1]+1) - log2(plotdata[, 2]+1),
  #State_ChildFC=Childstim-Childunstim)
  State_ChildFC = log2(plotdata[, 3]+1) - log2(plotdata[, 4]+1),
  #State=(y_Adultstim - y_Childstim) - (y_Adultunstim -y_Childunstim)
  State_FC = (log2(plotdata[, 1]+1) - log2(plotdata[, 3]+1)) - (log2(plotdata[, 2]+1)-log2(plotdata[, 4]+1)),
  adjP_state_age = plotdata$qvals.state.age) 

plotdata2<-plotdata1[,c(13,38:40,37)]

##Saving counts and metadata
write.table(plotdata1, file = "../data/glmmseq_Age_Monocytes_output.txt", sep = "\t", row.names = TRUE)
```
#**Results:** 
The stats slot key points: 
**i)**The corrected Akaike’s Information Criterion AIC, which can be used to compare different models. For each statistic, smaller values are desirable.
**ii)**Coefficients for each interaction show the percentage of variation in the response that is explained by the model. It is calculated as 1 minus the ratio of the error sum of squares (which is the variation that is not explained by model) to the total sum of squares (which is the total variation in the model).
**iii)**The estimated standard deviation of the error term, the lower the value of S, the better the conditional fitted equation describes the response at the selected factor settings. 
```{r, model fit statistics, echo=TRUE, warn.conflicts=F, quietly=T}
stats = summary(glmm_results)
kable(stats[order(stats[, 'state_bool:age_bool']), ]) %>%
  kable_styling() %>%
  scroll_box(width = "100%", height = "400px")
```
**The predictions** slot shows the estimated means based on each gene’s fitted model and their 95% confidence intervals
```{r, predicted values, echo=TRUE, warn.conflicts=F, quietly=T}
predict = data.frame(glmm_results@predict)
kable(predict) %>%
  kable_styling() %>%
  scroll_box(width = "100%", height = "400px")
```
```{r, calculated fc from model, echo=TRUE, warn.conflicts=F, quietly=T}
FC = summary(plotdata2)
kable(plotdata2[order(plotdata2[, 'qvals.state_bool.age_bool']), ]) %>%
  kable_styling() %>%
  scroll_box(width = "100%", height = "400px")
```

**Fold Change plots** calculated from the predictions slot.
Scatter plots comparing gene expression changes
between stimulated and unstimulated Monocytes cells in 10 paired samples from adults patients (n = 5) and children (n = 5).
Log2 fold change in expression for adults and children samples are represented on the x and y axis, respectively. Genes equally affected regardless off cmv state lie along the line of identity. Genes in green show significant (FDR < 0.05) overall change in expression after stimulation; those in blue/yellow show significantly differential change in expression over time between the two states based on significant (FDR < 0.05) interaction term state × age (line 67). Genes with greater absolute fold change for adults or children are shown in blue and yellow, respectively. **note** taken from https://doi.org/10.1038/s41591-022-01789-0 
```{r, Fold Change plots, include=TRUE, echo=TRUE}
#CXCL8, "FCGR1B", "IL1A", not found
monolabels = c("ADAM9", "ADAM10", "TLR1", "TLR2", "TLR4","TLR7", "TLR8", "CD163", "CD36", "CD93", "CR1", "FCAR", "FCER1G", "FCGR1A","FCGR2A", "FCGR3A", "CCL2", "CCL20", "CCL3", "CCL3L1", "CCL3L3", "CCL4", "CCL4L2", "CCL5", "CCL7", "CCR1", "CCRL2", "CX3CR1", "CXCL1", "CXCL16", "CXCL2", "CXCL3", "CXCR4","IL18", "IL1B", "IL6", "TNF", "S100A8","S100A9", "ISG15", "ISG20", "NFKB1", "NFKB2", "CD1C", "CD74", "CD83", "HLA-DPA1", "HLA-DPB1", "HLA-DQA1", "HLA-DQB1", "HLA-DRA", "HLA-DRB1", "HLA-DRB5")

fcPlot(glmm_results,
            x1var = "state_bool", 
            x2var = "age_bool", 
            graphics = "plotly",
            pCutoff = 0.05, useAdjusted = TRUE,
            labels = monolabels,
            fontSize = 12,
            labelFontSize = 4,
            colours = c('grey', 'green3', 'gold3', 'blue'))
```

**MA-plots** calculated from the predictions slot.
see line 57 for interpretation.
```{r, MA plots2, include=TRUE, echo=TRUE}
maPlots <- glmmSeq::maPlot(glmm_results,
                           x1var="state_bool",
                           x2var="age_bool",
                           x2Values=c("Adult", "Child"),
                           pCutoff=0.05,
                           colours=c('grey', 'green3', 'gold3', 'blue'),
                           labels=monolabels,
                           labelFontSize = 4,
                           useAdjusted = TRUE,
                           graphics="plotly")
maPlots$combined
```
